commit b8c15860719a3ca182b1e7689df231edc68899f5
Author: Adam Gołębiowski <adamg@pld-linux.org>
Date:   Sun Sep 23 16:11:21 2018 +0200

    ptstream.c: fix openssl-1.1 warnings

diff --git a/ptstream.c b/ptstream.c
index d1c5f44..bf3995f 100644
--- a/ptstream.c
+++ b/ptstream.c
@@ -208,7 +208,7 @@ int check_cert_names(X509 *cert, char *peer_host) {
 		for (i = 0; i < san_count; i++) {
 			gn = sk_GENERAL_NAME_value(gen_names, i);
 			if (gn->type == GEN_DNS && !(peer_host_is_ipv4 || peer_host_is_ipv6)) {
-				if (check_cert_valid_host((char*)ASN1_STRING_data(gn->d.ia5), peer_host)) {
+				if (check_cert_valid_host((char*)ASN1_STRING_get0_data(gn->d.ia5), peer_host)) {
 					return 1;
 				}
 			} else if (gn->type == GEN_IPADD) {
@@ -254,7 +254,7 @@ int stream_enable_ssl(PTSTREAM *pts, const char *proxy_arg) {
 	/* Initialise the connection */
 	SSLeay_add_ssl_algorithms();
 	if (args_info.enforcetls1_flag) {
-		meth = TLSv1_client_method();
+		meth = TLS_client_method();
 	} else {
 		meth = SSLv23_client_method();
 	}
